tests/publiccloud/sles4sap.pm

```diff
@@ -106,35 +119,53 @@ sub fence_node {
     $args{hostname} //= $self->{my_instance}->{instance_id};
 
     # Do the fencing!
-    record_info('Fencing info', 'Fencing done with systemctl poweroff');
-    $self->run_cmd(cmd => 'systemctl poweroff --force', timeout => 0);
-
-    # skip the one minute waiting
-    sleep 60;
-    my $start_time = time();
+    if (check_var('USE_FENCING', 'poweroff')) {
+        record_info('Fencing info', 'Fencing done with systemctl poweroff');
+        $self->run_cmd(cmd => 'systemctl poweroff --force', timeout => 0, quiet => 1);
+    } else {    # Default fencing uses CRM
+        record_info('Fencing info', 'Fencing done with crm');
+        $self->run_cmd(cmd => "crm -F node fence $args{hostname}", quiet => 1);
+    }
 
     # Wait till ssh disappear
-    while ((time() - $start_time) < $timeout) {
-        last unless (defined($self->{my_instance}->wait_for_ssh(timeout => 1, proceed_on_failure => 1)));
+    my $start_time = time;
+    while ((time - $start_time) < $timeout) {
+        last unless (defined($self->{my_instance}->wait_for_ssh(timeout => 1, proceed_on_failure => 1, quiet => 1)));
     }
-    my $fencing_time = time() - $start_time;
+    my $fencing_time = time - $start_time;
     die("Waiting for fencing failed!") unless ($fencing_time < $timeout);
     record_info('Fencing finished!', "Fencing done in ${fencing_time}s");
 
+    # We need to be sure that the server is stopped (if applicable)
+    if (check_var('PUBLIC_CLOUD_PROVIDER', 'EC2')) {
+        my $start_time = time;
+        while ($self->{my_instance}->get_state ne 'stopped') {
```

Comment:
why not to use https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/publiccloud/instance.pm#L269 ?

Reply:
in worst case if you think that this is some very special case which current implementation of `instance::stop` is not able to handle I will vote for extending current implementation of `instance::stop` with this logic and having some special flag which would trigger it instead of spreading stop logic into several places . 
@cfconrad  WDYT ?

Reply:
Not sure what is tested here. 
For me it looks like we are using `systemctl poweroff --force` or `crm -F node fence $args{hostname}` for power off and with this, he will not run the `aws ec2 stop-instances` command again.
So I agree with just checking the state.
But I'm wonder a bit about other providers like `gce` or `azure`.

Reply:
Actually, we try to simulate a cluster crash with `systemctl poweroff --force` or `crm -F node fence $args{hostname}`. 
We can not use `aws ec2 stop-instances` because it will do a clean shutdown I guess, that's why we need to check the status of the instance before starting it back. Maybe we can think about using a force shutdown command through the AWS cli if that command exists.

Reply:
The `stop` function is used to stop the VM, not to check the status. The VM is stopped by the cluster stack, and I need to check the status.

Reply:
> The `stop` function is used to stop the VM, not to check the status. The VM is stopped by the cluster stack, and I need to check the status.

no it does , see   https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/publiccloud/ec2.pm#L207
`stop` function is calling `provider::stop_instance` which doing first `aws ec2 stop-instances --instance-ids ' . $instance_id` which you obviously don't need here but later it making sure that instance is stopped by using almost the same loop like you introduced .

Reply:
On `azure` and `gce` when the node is fenced by the cluster stack it is automatically restarted but not on `ec2`. This is why we have to manually call the `start` function, but as the VM takes time to shutdown if we are not sure that the VM state is `stopped` we ended with an error when we call the `start` function.

This is why I implemented this loop only for `ec2`.

**Note:** the behavior of the STONITH agent (the "process" in charge of the automated stop/reboot) can be set to either `stop` or `reboot`, but it is configured differently depending on the provider recommendation (explanation from the devs, because I first asking for the same behavior for all providers!)

Reply:
> `stop` function is calling `provider::stop_instance` which doing first `aws ec2 stop-instances --instance-ids ' . $instance_id` which you obviously don't need here but later it making sure that instance is stopped by using almost the same loop like you introduced .

@asmorodskyi I remember to test to directly call `stop`, but as far as I remember if the state is already `stopped` the `ec2` command fails and thus the whole test. But I can try again if you want :-)